File: bot/src/web/api/session.ts
import type { NextFunction, Request, Response } from "express";
import { Router } from "express";
import { z } from "zod";
import type { AuditStore } from "../../data/audit.js";
import type { SessionStore } from "../../data/sessions.js";
import { SESSION_TTL_MS } from "../../data/sessions.js";
import type { UserStore } from "../../data/users.js";
import type { Logger } from "../../logger.js";
import {
  extractSessionToken,
  SESSION_COOKIE_NAME,
  validateSessionFromHeaders,
} from "../auth/validateSession.js";
import { parseWithSchema, zPassword, zUsername } from "../validate.js";

const FAILED_LOGIN_DELAY_MS = 250;

function setSessionCookie(res: Response, token: string): void {
  res.cookie(SESSION_COOKIE_NAME, token, {
    httpOnly: true,
    sameSite: "lax",
    secure: res.req.secure,
    path: "/",
    maxAge: SESSION_TTL_MS,
  });
}

function clearSessionCookie(res: Response): void {
  res.clearCookie(SESSION_COOKIE_NAME, { path: "/" });
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export function createSessionRouter(
  users: UserStore,
  sessions: SessionStore,
  audit: AuditStore,
  logger: Logger,
): Router {
  const router = Router();

  const requireAuthInline = (req: Request, res: Response, next: NextFunction) => {
    const result = validateSessionFromHeaders(req.headers.cookie, sessions);
    if (!result) {
      clearSessionCookie(res);
      res.status(401).json({ error: "unauthenticated" });
      return;
    }
    req.user = { id: result.userId, username: result.username, role: result.role };
    const token = extractSessionToken(req.headers.cookie);
    if (token) setSessionCookie(res, token);
    next();
  };

  router.get("/needs-setup", (_req, res) => {
    res.json({ needsSetup: users.countUsers() === 0 });
  });

  router.post("/setup", async (req, res) => {
    if (users.countUsers() !== 0) {
      res.status(409).json({ error: "already initialized" });
      return;
    }
    const parsed = parseWithSchema(
      z.object({ username: zUsername, password: zPassword }),
      req.body ?? {},
    );
    if (!parsed.ok) {
      res.status(400).json({ error: "invalid username or password" });
      return;
    }
    const { username, password } = parsed.data;
    try {
      const user = await users.createFirstUser(username, password);
      if (!user) {
        res.status(409).json({ error: "already initialized" });
        return;
      }
      const { token } = sessions.createSession(user.id);
      setSessionCookie(res, token);
      try {
        audit.record({
          actorId: user.id,
          actorUsername: user.username,
          targetUserId: user.id,
          targetUsername: user.username,
          action: "admin.first_created",
        });
      } catch (auditErr) {
        logger.warn({ err: auditErr, action: "admin.first_created" }, "audit insert failed");
      }
      logger.info({ userId: user.id, username }, "First admin created");
      res.json({ id: user.id, username: user.username, role: user.role });
    } catch (err) {
      logger.error({ err }, "setup failed");
      res.status(500).json({ error: "internal error", code: "INTERNAL_ERROR" });
    }
  });

  router.post("/login", async (req, res) => {
    const parsed = parseWithSchema(
      z.object({ username: z.string().min(1), password: z.string().min(1) }),
      req.body ?? {},
    );
    if (!parsed.ok) {
      res.status(400).json({ error: "invalid request" });
      return;
    }
    const { username, password } = parsed.data;
    const user = users.findByUsername(username);
    const ok = user ? await users.verifyPassword(password, user.passwordHash) : false;
    if (!user || !ok) {
      await delay(FAILED_LOGIN_DELAY_MS);
      res.status(401).json({ error: "invalid credentials" });
      return;
    }
    const { token } = sessions.createSession(user.id);
    setSessionCookie(res, token);
    res.json({ id: user.id, username: user.username, role: user.role });
  });

  router.post("/logout", (req, res) => {
    const token = extractSessionToken(req.headers.cookie);
    if (token) {
      sessions.deleteSession(token);
    }
    clearSessionCookie(res);
    res.status(204).end();
  });

  router.get("/me", requireAuthInline, (req, res) => {
    res.json(req.user);
  });

  router.post("/change-password", requireAuthInline, async (req, res) => {
    const { oldPassword, newPassword } = req.body ?? {};
    if (typeof oldPassword !== "string") {
      res.status(400).json({ error: "invalid request" });
      return;
    }
    const u = users.findById(req.user!.id);
    if (!u || !(await users.verifyPassword(oldPassword, u.passwordHash))) {
      await delay(FAILED_LOGIN_DELAY_MS);
      res.status(401).json({ error: "invalid credentials" });
      return;
    }
    const newParsed = parseWithSchema(zPassword, newPassword);
    if (!newParsed.ok) {
      res.status(400).json({ error: "invalid request" });
      return;
    }
    await users.changePassword(u.id, newParsed.data);
    const currentToken = extractSessionToken(req.headers.cookie);
    sessions.deleteAllForUser(u.id, currentToken ?? undefined);
    try {
      audit.record({
        actorId: u.id,
        actorUsername: u.username,
        targetUserId: u.id,
        targetUsername: u.username,
        action: "user.password_changed",
      });
    } catch (auditErr) {
      logger.warn({ err: auditErr, action: "user.password_changed" }, "audit insert failed");
    }
    res.status(204).end();
  });

  return router;
}

File: crates/mp-http/src/session.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

//! Session cookie API — `/api/session/*`. First-account-is-admin.

use axum::body::Body;
use axum::extract::State;
use axum::http::{header, Request, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde::Deserialize;
use serde_json::json;
use tokio::time::{sleep, Duration};

use crate::{clear_session_cookie, set_session_cookie, AppState};

pub use mp_config::SESSION_COOKIE_NAME;

const FAILED_LOGIN_DELAY_MS: u64 = 250;

#[derive(Deserialize)]
pub struct SetupBody {
    username: String,
    password: String,
}

#[derive(Deserialize)]
pub struct LoginBody {
    username: String,
    password: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChangePasswordBody {
    #[serde(alias = "oldPassword")]
    old_password: Option<String>,
    #[serde(alias = "currentPassword")]
    current_password: Option<String>,
    new_password: String,
}

pub async fn needs_setup(State(st): State<AppState>) -> impl IntoResponse {
    match st.db.user_count() {
        Ok(n) => Json(json!({ "needsSetup": n == 0 })).into_response(),
        Err(e) => {
            tracing::error!(error = %e, "needs-setup");
            (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({"error":"internal error"})))
                .into_response()
        }
    }
}

pub async fn setup(State(st): State<AppState>, req: Request<Body>) -> Response {
    if let Err(wait) = st.setup_limit.take(&client_ip(&req, &st)) {
        return too_many(wait);
    }
    let secure = is_secure(&req);
    let body = match read_json::<SetupBody>(req).await {
        Ok(b) => b,
        Err(r) => return r,
    };
    if !valid_username(&body.username) || !valid_password(&body.password) {
        return (StatusCode::BAD_REQUEST, Json(json!({"error":"invalid username or password"})))
            .into_response();
    }
    match st.db.user_count() {
        Ok(0) => {}
        Ok(_) => {
            return (StatusCode::CONFLICT, Json(json!({"error":"already initialized"})))
                .into_response();
        }
        Err(e) => {
            tracing::error!(error = %e, "setup count");
            return internal();
        }
    }
    match st.db.users().create_first_user(&body.username, &body.password) {
        Ok(Some(user)) => match st.db.sessions().create_session(&user.id) {
            Ok((token, _)) => {
                let body = json!({
                    "id": user.id,
                    "username": user.username,
                    "role": user.role.as_str(),
                });
                let res = (StatusCode::OK, Json(body)).into_response();
                set_session_cookie(res, &token, secure)
            }
            Err(e) => {
                tracing::error!(error = %e, "setup session");
                internal()
            }
        },
        Ok(None) => (StatusCode::CONFLICT, Json(json!({"error":"already initialized"})))
            .into_response(),
        Err(e) => {
            tracing::error!(error = %e, "setup");
            internal()
        }
    }
}

pub async fn login(State(st): State<AppState>, req: Request<Body>) -> Response {
    if let Err(wait) = st.login_limit.take(&client_ip(&req, &st)) {
        return too_many(wait);
    }
    let secure = is_secure(&req);
    let body = match read_json::<LoginBody>(req).await {
        Ok(b) => b,
        Err(r) => return r,
    };
    if body.username.is_empty() || body.password.is_empty() {
        return (StatusCode::BAD_REQUEST, Json(json!({"error":"invalid request"}))).into_response();
    }
    let user = match st.db.users().find_by_username(&body.username) {
        Ok(u) => u,
        Err(e) => {
            tracing::error!(error = %e, "login lookup");
            return internal();
        }
    };
    let ok = match &user {
        Some(u) => st
            .db
            .users()
            .verify_password(&body.password, &u.password_hash)
            .unwrap_or(false),
        None => false,
    };
    if !ok {
        sleep(Duration::from_millis(FAILED_LOGIN_DELAY_MS)).await;
        return (StatusCode::UNAUTHORIZED, Json(json!({"error":"invalid credentials"})))
            .into_response();
    }
    let user = user.unwrap();
    match st.db.sessions().create_session(&user.id) {
        Ok((token, _)) => {
            let body = json!({
                "id": user.id,
                "username": user.username,
                "role": user.role.as_str(),
            });
            let res = (StatusCode::OK, Json(body)).into_response();
            set_session_cookie(res, &token, secure)
        }
        Err(e) => {
            tracing::error!(error = %e, "login session");
            internal()
        }
    }
}

pub async fn logout(State(st): State<AppState>, req: Request<Body>) -> Response {
    if let Some(token) = extract_session_token(req.headers().get(header::COOKIE)) {
        let _ = st.db.sessions().delete_session(&token);
    }
    let res = StatusCode::NO_CONTENT.into_response();
    clear_session_cookie(res)
}

pub async fn me(State(st): State<AppState>, req: Request<Body>) -> Response {
    match current_user(&st, req.headers().get(header::COOKIE)) {
        Some(u) => Json(json!({
            "id": u.user_id,
            "username": u.username,
            "role": u.role.as_str(),
        }))
        .into_response(),
        None => {
            let res = (
                StatusCode::UNAUTHORIZED,
                Json(json!({"error":"unauthenticated"})),
            )
                .into_response();
            clear_session_cookie(res)
        }
    }
}

pub async fn change_password(State(st): State<AppState>, req: Request<Body>) -> Response {
    let cookie = req.headers().get(header::COOKIE).cloned();
    let Some(sess) = current_user(&st, cookie.as_ref()) else {
        return (
            StatusCode::UNAUTHORIZED,
            Json(json!({"error":"unauthenticated"})),
        )
            .into_response();
    };
    let body = match read_json::<ChangePasswordBody>(req).await {
        Ok(b) => b,
        Err(r) => return r,
    };
    let old = body
        .old_password
        .or(body.current_password)
        .unwrap_or_default();
    let Some(u) = st.db.users().find_by_id(&sess.user_id).ok().flatten() else {
        return (StatusCode::UNAUTHORIZED, Json(json!({"error":"invalid credentials"})))
            .into_response();
    };
    let ok = st
        .db
        .users()
        .verify_password(&old, &u.password_hash)
        .unwrap_or(false);
    if !ok {
        sleep(Duration::from_millis(FAILED_LOGIN_DELAY_MS)).await;
        return (StatusCode::UNAUTHORIZED, Json(json!({"error":"invalid credentials"})))
            .into_response();
    }
    if !valid_password(&body.new_password) {
        return (StatusCode::BAD_REQUEST, Json(json!({"error":"invalid request"}))).into_response();
    }
    if let Err(e) = st.db.users().change_password(&u.id, &body.new_password) {
        tracing::error!(error = %e, "change-password");
        return internal();
    }
    let current = extract_session_token(cookie.as_ref());
    let _ = st
        .db
        .sessions()
        .delete_all_for_user(&u.id, current.as_deref());
    StatusCode::NO_CONTENT.into_response()
}

pub fn current_user(
    st: &AppState,
    cookie: Option<&axum::http::HeaderValue>,
) -> Option<mp_db::SessionValidation> {
    let token = extract_session_token(cookie)?;
    st.db.sessions().validate_and_touch(&token).ok().flatten()
}

pub fn extract_session_token(header: Option<&axum::http::HeaderValue>) -> Option<String> {
    let raw = header?.to_str().ok()?;
    for part in raw.split(';') {
        let part = part.trim();
        if let Some(v) = part.strip_prefix(&format!("{SESSION_COOKIE_NAME}=")) {
            if !v.is_empty() {
                return Some(v.to_string());
            }
        }
    }
    None
}

fn valid_username(s: &str) -> bool {
    let n = s.len();
    (3..=32).contains(&n)
        && s.chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '_' | '-' | '.'))
}

fn valid_password(s: &str) -> bool {
    (8..=200).contains(&s.len())
}

fn is_secure(req: &Request<Body>) -> bool {
    req.headers()
        .get("x-forwarded-proto")
        .and_then(|v| v.to_str().ok())
        .is_some_and(|v| v.eq_ignore_ascii_case("https"))
}

fn client_ip(req: &Request<Body>, st: &AppState) -> String {
    if st.config.trust_proxy {
        if let Some(xff) = req
            .headers()
            .get("x-forwarded-for")
            .and_then(|v| v.to_str().ok())
        {
            let hops = st.config.trust_proxy_hops.max(1) as usize;
            let parts: Vec<&str> = xff.split(',').map(|s| s.trim()).collect();
            if !parts.is_empty() {
                let idx = parts.len().saturating_sub(hops);
                return parts[idx].to_string();
            }
        }
    }
    req.headers()
        .get("x-real-ip")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("unknown")
        .to_string()
}

async fn read_json<T: serde::de::DeserializeOwned>(req: Request<Body>) -> Result<T, Response> {
    let bytes = axum::body::to_bytes(req.into_body(), 100 * 1024)
        .await
        .map_err(|_| {
            (
                StatusCode::BAD_REQUEST,
                Json(json!({"error":"invalid request"})),
            )
                .into_response()
        })?;
    serde_json::from_slice(&bytes).map_err(|_| {
        (
            StatusCode::BAD_REQUEST,
            Json(json!({"error":"invalid request"})),
        )
            .into_response()
    })
}

fn too_many(wait: u64) -> Response {
    (
        StatusCode::TOO_MANY_REQUESTS,
        [(header::RETRY_AFTER, wait.to_string())],
        Json(json!({"error":"rate limited"})),
    )
        .into_response()
}

fn internal() -> Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({"error":"internal error","code":"INTERNAL_ERROR"})),
    )
        .into_response()
}
